File: API_java/app/src/main/java/com/example/api_java/MainActivity.java
package com.example.api_java;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;

import com.google.gson.Gson;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class MainActivity extends AppCompatActivity {
    private Button btn_quary;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        btn_quary = findViewById(R.id.btn_query);
        registerReceiver(receiver,new IntentFilter("Mymessage"));
        btn_quary.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Request req = new Request.Builder()
                        .url("https://data.taipei/opendata/datalist/apiAccess?" +
                                "scope=resourceAquire&rid=55ec6d6e-dc5c-4268-a725-d04cc262172b").build();

                new OkHttpClient().newCall(req).enqueue(new Callback() {
                    @Override
                    public void onFailure(@NotNull Call call, @NotNull IOException e) {
                        Log.e("查詢失敗",e.toString());
                    }

                    @Override
                    public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {
                        sendBroadcast(new Intent("Mymessage")
                                .putExtra("json",response.body().string()));
                    }
                });
            }
        });
    }


    private BroadcastReceiver receiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            Data data = new Gson().fromJson(intent.getExtras().getString("json"),Data.class);
            final String[] items = new String[data.result.results.length];
            for(int i=0;i<items.length;i++){
                items[i] = "\n列車即將進入 :"+
                        data.result.results[i].Station+
                        "\n列車行駛目的地 :"+
                        data.result.results[i].Destination;

            }
            MainActivity.this.runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    new AlertDialog.Builder(MainActivity.this)
                            .setTitle("台北捷運列車到站站名")
                            .setItems(items,null)
                            .show();
                }
            });
        }
    };

}

class Data {
    Result result;
    class Result{
        Results[] results;
        class  Results{
            String Station;
            String Destination;
        }
    }
}

File: Broadcast_kotlin/app/src/main/java/com/example/broadcast_kotlin/MainActivity.kt
package com.example.broadcast_kotlin

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.widget.Toast
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {
    companion object{
        var flag: Boolean = false
    }
    private val receiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val b = intent.extras
            tv_clock.text=
                String.format(
                    "%02d:%02d:%02d",
                    b!!.getInt("H"), b.getInt("M"), b.getInt("S")
                )
        }
    }
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        registerReceiver(receiver, IntentFilter("MyMessage"))
        flag = MyService.flag
        if (flag)
            btn_start.text = "暫停"
        else
            btn_start.text = "開始"

        btn_start.setOnClickListener {
            flag = !flag
            if (flag) {
                btn_start.text = "暫停"
                Toast.makeText(
                    this@MainActivity, "計時開始",
                    Toast.LENGTH_SHORT
                ).show()
            } else {
                btn_start.text = "開始"
                Toast.makeText(
                    this@MainActivity, "計時暫停",
                    Toast.LENGTH_SHORT
                ).show()
            }
            startService(Intent(this@MainActivity, MyService::class.java).putExtra("flag", flag))
        }

    }
    public override fun onDestroy() {
        super.onDestroy()
        unregisterReceiver(receiver)
    }
}
